caddyconfig/httpcaddyfile/directives.go

```diff
@@ -502,7 +510,7 @@ type (
 	// tokens from a global option. It is passed the tokens to parse and
 	// existing value from the previous instance of this global option
 	// (if any). It returns the value to associate with this global option.
-	UnmarshalGlobalFunc func(d *caddyfile.Dispenser, existingVal interface{}) (interface{}, error)
+	UnmarshalGlobalFunc func(h GlobalHelper, existingVal interface{}) (interface{}, error)
```

Comment:
I understand the purpose of this change, but this is technically a breaking change. This is meant to be available to third party plugins so they can register global options of their own.

That said, I don't think any actually do yet because support for it was incomplete (see https://github.com/caddyserver/caddy/issues/3634) but that was resolved in https://github.com/caddyserver/caddy/pull/3990 which will be released in v2.4.0.

So if we need to make this change, we'll have to make the change before the v2.4.0 release to avoid future breakages.

I'm thinking this refactor of `UnmarshalGlobalFunc` might need to also be moved into a separate PR and merged ahead of this one, but I'll let @mholt chime in first.

Reply:
Ah! Great point.

Without this or something like it, it might be a bit hacky to do the check for duplicate/colliding log configurations, and it would be something that had to be in Caddy itself, i.e. not possible as part of the public API. FWIW, this seems like it may be a beneficial construct to have going forward to make the API more easily extensible in a non-breaking way, like directives work today.

Reply:
Hey @mholt @francislavoie , just wanted to check in here, any further thoughts on whether or not this is an acceptable change to make?

If there is an issue with this interface shift let me know and I take another look here.

Reply:
It hasn't been released yet, so we can technically change it, though that's not really desirable.

I don't _love_ the `GlobalHelper` state (we do already have something similar in `Helper`: https://pkg.go.dev/github.com/caddyserver/caddy/v2/caddyconfig/httpcaddyfile#Helper)

If the only goal is to pass in the state, then maybe that could/should just be another argument to `UnmarsahlGlobalFunc` functions.

Reply:
Looking at this again, I'm actually able to do what's needed without the state involved by using a healthy dose of type assertions to look at the "existing values" passed into the UnmarshalGlobalFunc. I've pushed commits for that now that I can squash down if the result looks amenable.

Reply:
I went ahead and rebased on master and squashed things down